### ISSUE TITLE
(I)Python Client

### ISSUE THREAD

**Issue by ax3l** [ComputationalRadiationPhysics/isaac#40]
We should write an additional client class that can be used together with [widgets](https://github.com/ipython/ipywidgets) in Jupyther (IPython) notebooks.

**Comment by bussmann** [comment]
+1

**Comment by theZiz** [comment]
Yeah, sure.

As you may know, I just wrote another simple client, too. ISAAC itself doesn't know any clients, but protocols. Atm it knows websocket and since yesterday tcp. So just tell me, how the protocol for the python widget shall lock like and I do my part.

Anyway, you can already connect with python via tcp to isaac and get json messages if you want. So maybe I don't have to implement anything. :D

**Comment by bussmann** [comment]
This is not yet another client. It is EXTREMELY important to pythonize the PIConGPU workflow and make it useable for non-experts. ISAAC is an important part of that workflow.

**Comment by bussmann** [comment]
P.S.: I love that it now speaks websockets&TCP. Any  plans for WebRTC?

**Comment by theZiz** [comment]
You got me wrong or better: I was unspecific. The isaac **server** just needs a protocol to pass json messages to the insitu library. Python should be able to use tcp for this. However you are right, the **library** should probably get some new fancy features for the python binding. ;) But the whole API will get some major changes anyway when my WHK employment starts...

About WebRTC: No plans atm. Tbh I am not a big fan of WebRTC. Too much packed in one little  very problem specific API. E.g. WebRTC can handle h264 rtp streams - but the video tag cannot! Imho they should focus on basic html5 features instead of adding more plugins directly to the browser. Enough rant. However the API and the needed environment is still very complex. At least it cannot be added just by the way. Before I can even _decice_, whether WebRTC is useful and doable I need to read more about it, maybe even play a bit with it.

**Comment by bussmann** [comment]
Fair enough, I like a step by step approach.

**Comment by bussmann** [comment]
...and just to be a nuisance: It might not seems necessary, but looking into using GrayBat might be a good thing on your bucket list, too. *duckandcover*

**Comment by theZiz** [comment]
I'll try outsourcing this to @erikzenker. :>

**Comment by erikzenker** [comment]
Hot plugging is underway ;-)  ...and as I told you yesterday, there are plans for graybat.io

**Comment by erikzenker** [comment]
...and now there exist plans for PyBat

**Comment by ax3l** [comment]
> As you may know, I just wrote another simple client, too. ISAAC itself doesn't know any clients, but protocols. Atm it knows websocket and since yesterday tcp. So just tell me, how the protocol for the python widget shall lock like and I do my part.

yes, could e.g, be an other project labeled `isaac-python` because making up names that are super unrelated helps nobody ;)

> It is EXTREMELY important to pythonize the PIConGPU workflow and make it useable for non-experts. ISAAC is an important part of that workflow.

absolutely. but to be fair: it is just an other client and I love that it is so simple.

> WebRTC

off-topic guys, open additional issues please ;)
we should also document why webrtc was not implemented, we discussed a lot offline at the beginning of your master's @theZiz , maybe you want to put them in an issue for later reference what to do (or why not to do it).

> Fair enough, I like a step by step approach

exactly

> graybat.io

fantastic and yet: an other issue please :)

**Comment by ax3l** [comment]
I did split the thread, sorry guys - moderated too many forums in the past. Could not resist.

**Comment by bussmann** [comment]
I thought of opening different threads but was lazy, so I apologize :-)

**Comment by theZiz** [comment]
No... Not more issues I can't close for the next months.
Micha will never let me free out of the HZDR basement if this continues!

**Comment by bussmann** [comment]
gnihihihihiMUAHAHAHAHAAAA!

**Comment by theZiz** [comment]
Please Master... Dobby wants to be free!

(SCNR. Back to serious discussions about IPython right now! *cough*)

**Comment by ax3l** [comment]
haha, this thread will never clean up.

here is a troll face :trollface: , you all earned it :dancers:

**Comment by ax3l** [comment]
Once could lend some of the GUI elements from [ipyvolume](https://github.com/maartenbreddels/ipyvolume) (steering, controls) and add the output stream as in the HTML client.

**Comment by ax3l** [comment]
first draft commited to https://github.com/ComputationalRadiationPhysics/isaac-python
